presto-iceberg/src/main/java/com/facebook/presto/iceberg/procedure/RemoveOrphanFiles.java

```diff
@@ -0,0 +1,232 @@
+
+        // Remove unused metadata and data files older than 3 days by default
+        // This default value is consistent with Spark procedure `remove_orphan_files` on Iceberg, see:
+        //  https://iceberg.apache.org/docs/1.5.2/spark-procedures/#remove_orphan_files
+        long expiration = olderThan == null ?
+                System.currentTimeMillis() - DAYS.toMillis(REMOVE_UNUSED_FILES_OLDER_THAN_IN_DAYS) :
+                olderThan.isLegacyTimestamp() ? olderThan.getMillisUtc() : olderThan.getMillis();
+        scanAndDeleteInvalidFiles(icebergTable, clientSession, schemaTableName, expiration, validDataFileNames.build(), dataLocation(icebergTable));
+        scanAndDeleteInvalidFiles(icebergTable, clientSession, schemaTableName, expiration, validMetadataFileNames.build(), metadataLocation(icebergTable));
+    }
+
+    private void scanAndDeleteInvalidFiles(Table table, ConnectorSession session, SchemaTableName schemaTableName, long expiration, Set<String> validFiles, String folderFullPath)
+    {
+        try {
+            List<String> filesToDelete = new ArrayList<>();
+            FileSystem fileSystem = getFileSystem(session, this.hdfsEnvironment, schemaTableName, new Path(table.location()));
+            Path fullPath = new Path(folderFullPath);
+            if (!fileSystem.exists(fullPath)) {
+                return;
+            }
+            RemoteIterator<LocatedFileStatus> allFiles = fileSystem.listFiles(fullPath, true);
+            while (allFiles.hasNext()) {
+                LocatedFileStatus entry = allFiles.next();
+                if (entry.getModificationTime() <= expiration && !validFiles.contains(entry.getPath().getName())) {
+                    filesToDelete.add(entry.getPath().toString());
+                    if (filesToDelete.size() >= BATCH_DELETE_FILES_COUNT) {
+                        LOGGER.debug("Deleting files while removing orphan files for table %s : %s", schemaTableName, filesToDelete);
+                        CatalogUtil.deleteFiles(table.io(), filesToDelete, folderFullPath, true);
+                        filesToDelete.clear();
+                    }
+                }
+                else {
+                    LOGGER.debug("%s retained while removing orphan files %s", entry.getPath().toString(), schemaTableName.getTableName());
+                }
+            }
+
+            if (!filesToDelete.isEmpty()) {
+                LOGGER.debug("Deleting files while removing orphan files for table %s : %s", schemaTableName, filesToDelete);
+                CatalogUtil.deleteFiles(table.io(), filesToDelete, folderFullPath, true);
+            }
+        }
+        catch (IOException e) {
+            throw new PrestoException(ICEBERG_FILESYSTEM_ERROR, "Failed accessing data for table: " + schemaTableName, e);
+        }
+    }
+
+    private String extractFileName(String path)
+    {
+        return path.substring(path.lastIndexOf('/') + 1);
+    }
```

Comment:
In some rare circumstances, the authority/scheme of data files may not equal between table metadata and the real storage system. Referring to the comment for `DeleteOrphanFiles.prefixMismatchMode(...)` and test cases `testRemoveOrphanFilesProcedureWithPrefixMode()` and `testRemoveOrphanFilesProcedureWithEqualAuthorities()` in `TestRemoveOrphanFilesProcedure`.

So we use a file name rather than the fully qualified path of the file to choose an `IGNORE` strategy, that is, we do not delete the files that has mismatches in authority/scheme, which is more safer than the `DELETE` strategy.

Besides, this method has exactly the same implementation logic as `org.apache.hadoop.fs.Path.getName()`. The file paths maintained in Iceberg table metadata are all fully qualified paths, which are suitable for constructing a Hadoop Path (see the comment in `ContentFile.path()` and `ManifestFile.path()`). That means, we can omit the verification of these paths (which would be applied when create a new `Path` instance) and just get their names.

Reply:
Makes sense, thanks for the explanation